projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/control/LLVMLoopDispatchNode.java

```diff
@@ -0,0 +1,265 @@
+                        basicBlockIndex = successors[i];
+                        nullDeadSlots(frame, basicBlockIndex, beforeBlockNuller);
+                        continue outer;
+                    }
+                }
+
+                int i = successors.length - 1;
+                assert successorBasicBlockIndex == successors[i];
+                if (CompilerDirectives.inInterpreter()) {
+                    bb.increaseBranchProbability(i);
+                }
+                executePhis(frame, indirectBranchNode, i);
+                nullDeadSlots(frame, basicBlockIndex, afterBlockNuller);
+                basicBlockIndex = successors[i];
+                nullDeadSlots(frame, basicBlockIndex, beforeBlockNuller);
+                continue outer;
+            } else if (controlFlowNode instanceof LLVMBrUnconditionalNode) {
+                LLVMBrUnconditionalNode unconditionalNode = (LLVMBrUnconditionalNode) controlFlowNode;
+                unconditionalNode.execute(frame); // required for instrumentation
+                executePhis(frame, unconditionalNode, 0);
+                nullDeadSlots(frame, basicBlockIndex, afterBlockNuller);
+                basicBlockIndex = unconditionalNode.getSuccessor();
+                nullDeadSlots(frame, basicBlockIndex, beforeBlockNuller);
+                continue outer;
+            } else if (controlFlowNode instanceof LLVMInvokeNode) {
+                LLVMInvokeNode invokeNode = (LLVMInvokeNode) controlFlowNode;
+                try {
+                    invokeNode.execute(frame);
+                    executePhis(frame, invokeNode, LLVMInvokeNode.NORMAL_SUCCESSOR);
+                    nullDeadSlots(frame, basicBlockIndex, afterBlockNuller);
+                    basicBlockIndex = invokeNode.getNormalSuccessor();
+                    nullDeadSlots(frame, basicBlockIndex, beforeBlockNuller);
+                    continue outer;
+                } catch (LLVMUserException e) {
+                    frame.setObject(exceptionValueSlot, e);
+                    executePhis(frame, invokeNode, LLVMInvokeNode.UNWIND_SUCCESSOR);
+                    nullDeadSlots(frame, basicBlockIndex, afterBlockNuller);
+                    basicBlockIndex = invokeNode.getUnwindSuccessor();
+                    nullDeadSlots(frame, basicBlockIndex, beforeBlockNuller);
+                    continue outer;
+                }
+            } else {    // some control flow nodes should be never part of a loop
+                CompilerAsserts.neverPartOfCompilation();
+                throw new UnsupportedOperationException("unexpected controlFlowNode type: " + controlFlowNode);
+            }
+        }
+
+        if(basicBlockIndex != headerId) {
+            frame.setInt(successorSlot, basicBlockIndex);
+            return false;
+        }
+
+        return true;
+    }
+
+    @ExplodeLoop
+    private static void executePhis(VirtualFrame frame, LLVMControlFlowNode controlFlowNode, int successorIndex) {
```

Comment:
share the helper functions that are down here with the LLVMDispatchBasicBlockNode (this will not impact PE)